### ISSUE TITLE
Limit-events and truncating of data

### ISSUE THREAD

**Issue by LinusU** [expressjs/multer#128]
I've been thinking a lot on all of the different limits and events that multer can emit. Specifically, I can't find a use case where you would want to handle truncated data, and I actually think that it would be better if we removed all the the `on*Limit` events and instead called `next` with an `new Error` that has an appropriate code set.

E.g. too large file could be `err.code = 'LIMIT_FILE_SIZE'`.

I also think that we should remove the `onError` event in favour for just passing the errors to `next`.

If anyone would want to handle the errors only from multer (instead of passing them to their global error handler), they could add an error handler right after, like so:

```javascript
app.use(multer({ ... }))
app.use(function (err, req, res, next) {
  res.end('Multer error :(')
})
```

I think that the best argument for this is that it vastly improves out-of-the-box usage, especially for beginners. We currently don't store any information regarding truncated field names and values, which makes it a pain to debug when the value is cut of. It's also not very feasible to check `file.truncated` on every file.

I believe that the module should be as easy as this to use:

```javascript
app.use(multer({ dest: 'uploads/' }))
```

For that to work, it's important that we don't silently ignore errors.

**Comment by jpfluger** [comment]
What you suggest should work as an "all-or-nothing" approach to error handling. I think "all-or-nothing" needs to be implemented in some facet, so go for it!  

But in addition, is there a way to be more detailed, to detect and handle errors on a per-file basis?  Example: client uploads 5 files of varying sizes. Of these five 5 files, some upload just find and others don't (or are hung). The next() is called once all five files have completely uploaded or erred (timeout?).  If we can achieve detail control over each file, then bake-in support for user-controlled abort, resume, be informed about the status of uploads.  

How might this look? I don't think multer should do the heavy lifting here. Maybe a suite of EMITs allowing independent modules to build in user-controlled support?  (eg via socket.io)  ...which dovetails into your suggestion in https://github.com/expressjs/multer/issues/124.

**Comment by jpfluger** [comment]
Yes, error codes are good. I had an implementation for this last year but didn't want to rattle the boat much, as I was new to multer. Go for it! :+1:

**Comment by LinusU** [comment]
It would be really cool to have per file error handling, not sure on how to do it whilst still providing a great experience for our developers thought.

PHP sets `error` on each file but than that requires checking each file for errors again. Maybe we could introduce an option that dosen't send the error to next but instead sets it on the file in question.

**Comment by LinusU** [comment]
I just realized that it's way more easier to catch errors just from multer:

```javascript
var mw = multer({ ... })

mw(req, res, function (err) {
  // handle err here
})
```

**Comment by LinusU** [comment]
This is now fixed in `1.0.0`
